### ISSUE TITLE
Experimental Strict Type Guard unexpected Never

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#6134]
**Describe the bug**
```python
from collections.abc import Mapping

from typing_extensions import TypeGuard, TypeVar

T = TypeVar("T")


def is_type_dict(d: Mapping[str, object], typ: type[T]) -> TypeGuard[dict[str, T]]:
    ...


def foo(x: dict[str, object]):
    assert is_type_dict(x, int)

    reveal_type(x)  # Expected: dict[str, int]
```

Reveal type shows Never when `"enableExperimentalFeatures": true`. It shows `dict[str, int]` as expected when "enableExperimentalFeatures": false.

**VS Code extension or command-line**
Latest pylance version `v2023.10.21`.

Small thing I noticed when I was curious from PEP discussion about effects of enabling StrictTypeGuard. I see total 9 error changes where 6 are good (unnecessary type ignore), an error similar to this, and last 2 are unexpected. The last 2 look correct errors to me, but not sure why strictness would matter and will open a separate issue for.

edit: The last 2 errors I simplified and looked at more. I understand now how strictness can make a difference and error looks correct to me. So happy to see for ~30k lines of code only 1 error with strict typeguard looks false positive.

**Comment by erictraut** [comment]
The behavior is correct. Under PEP 724, this type guard function is considered "strict" because `dict[str, T]` is a subtype of `Mapping[str, object]`.

When strict narrowing is applied, it follows proper type theory to produce the results. The resulting narrowed type is an intersection between the input type `A` and the type guard return type `R` — or some approximation of this intersection. In your example, `A` is `dict[str, object]` and `R` is `dict[str, int]`. The intersection between these two types is `Never` because the type parameter for `dict` is invariant.

This makes sense if you think about it. For type-safe narrowing, the narrowed type must be a subtype of the pre-narrowed type. If you start with a `dict[str, object]`, you cannot safely narrow it to `dict[str, int]`.

**Comment by hmc-cs-mdrissi** [comment]
Hmm a bit confused. Here's few other variations,

```
_T = TypeVar("_T")


def is_type_dict1(d: Mapping[str, _T], typ: type[T]) -> TypeGuard[dict[str, T]]:
    ...

def is_type_dict2(d: Mapping[str, Any], typ: type[T]) -> TypeGuard[dict[str, T]]:
    ...
```

is_type_dict1 produces dict[str, int] when used for `assert is_type_dict(x, int)`. is_type_dict2 leads to Never. Is that also expected? I tend to think of Mapping[str, _T]/Mapping[str, Any] as same when _T is only used once.

Your solution works. Still a bit weird to me that this code,

```python
def is_type_dict(d: dict[str, object], typ: type[T]) -> TypeGuard[dict[str, T]]:
    ...

def foo(x: Mapping[str, object]):
  assert isinstance(x, dict) # Narrows x to dict[str, object]
  assert is_type_dict(x) # Narrows x to dict[str, int]
```

is fine, but combining the two asserts into one typeguard of `def is_type_dict(d: Mapping[str, object], typ: type[T]) -> TypeGuard[dict[str, T]]:` is not fine. Implementation of is_type_dict was isinstance(d, dict) + check each value is of type[T].
